kratos/utilities/nodal_sensitivity_builder.h

```diff
@@ -0,0 +1,288 @@
+    }
+
+    template <class TDataType>
+    void BuildNodalSolutionStepConditionContributions(Variable<TDataType> const& rVariable,
+                                                      double ScalingFactor)
+    {
+        KRATOS_TRY;
+
+        auto& r_conditions = mrModelPart.Conditions();
+        const auto& r_process_info = mrModelPart.GetProcessInfo();
+
+        const int num_threads = OpenMPUtils::GetNumThreads();
+        std::vector<Vector> local_sensitivity(num_threads);
+        std::vector<Vector> partial_sensitivity(num_threads);
+        std::vector<Vector> adjoint_vector(num_threads);
+        std::vector<Matrix> sensitivity_matrix(num_threads);
+
+#pragma omp parallel
+        {
+            ModelPart::ConditionIterator conditions_begin;
+            ModelPart::ConditionIterator conditions_end;
+            OpenMPUtils::PartitionedIterators(r_conditions, conditions_begin, conditions_end);
+            int k = OpenMPUtils::ThisThread();
+
+            for (auto it = conditions_begin; it != conditions_end; ++it)
+            {
+                Condition::GeometryType& r_geom = it->GetGeometry();
+
+                it->CalculateSensitivityMatrix(rVariable, sensitivity_matrix[k], r_process_info);
+
+                if (sensitivity_matrix[k].size1() == 0 || HasActiveNodes(r_geom) == false)
+                    continue;
+
+                mpResponseFunction->CalculatePartialSensitivity(
+                    *it, rVariable, sensitivity_matrix[k],
+                    partial_sensitivity[k], r_process_info);
+
+                it->GetValuesVector(adjoint_vector[k]);
+
+                if (local_sensitivity[k].size() != sensitivity_matrix[k].size1())
+                    local_sensitivity[k].resize(sensitivity_matrix[k].size1(), false);
+
+                noalias(local_sensitivity[k]) =
+                    ScalingFactor * (prod(sensitivity_matrix[k], adjoint_vector[k]) +
+                                     partial_sensitivity[k]);
+
+                AssembleNodalSolutionStepSensitivityContribution(
+                    rVariable, local_sensitivity[k], r_geom);
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    bool HasActiveNodes(GeometryType const& rGeom)
+    {
+        bool result = false;
+        for (unsigned i_node = 0; i_node < rGeom.PointsNumber(); ++i_node)
+            if (rGeom[i_node].GetValue(UPDATE_SENSITIVITIES))
+            {
+                result = true;
+                break;
+            }
+        return result;
+    }
+
+    void AssembleNodalSolutionStepSensitivityContribution(Variable<double> const& rVariable,
+                                                          Vector const& rSensitivityVector,
+                                                          GeometryType& rGeom)
+    {
+        unsigned int index = 0;
+        for (unsigned int i_node = 0; i_node < rGeom.PointsNumber(); ++i_node)
+        {
+            if (rGeom[i_node].GetValue(UPDATE_SENSITIVITIES))
```

Comment:
accessing a `const&` to the node would make this threadsafe if `UPDATE_SENSITIVITIES` is not set.